connectors/sources/servicenow.py

```diff
@@ -0,0 +1,346 @@
+        self.services = self.configuration["services"]
+        self.retry_count = self.configuration["retry_count"]
+
+    def _get_session(self):
+        """Generate aiohttp client session with configuration fields.
+
+        Returns:
+            aiohttp.ClientSession: An instance of Client Session
+        """
+
+        if self.session:
+            return self.session
+
+        logger.debug("Generating aiohttp client session")
+        connector = aiohttp.TCPConnector(limit=MAX_CONCURRENCY_SUPPORT)
+        basic_auth = aiohttp.BasicAuth(
+            login=self.configuration["username"],
+            password=self.configuration["password"],
+        )
+        request_headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+        timeout = aiohttp.ClientTimeout(total=None)  # pyright: ignore
+
+        self.session = aiohttp.ClientSession(
+            connector=connector,
+            base_url=self.configuration["url"],
+            auth=basic_auth,
+            headers=request_headers,
+            timeout=timeout,
+            raise_for_status=True,
+        )
+        return self.session
+
+    async def _api_call(self, url, params, is_attachment=False):
+        """Handle every api call to ServiceNow instance with retries.
+
+        Args:
+            url (str): ServiceNow query to be executed.
+            params (dict): Params to format the query.
+            is_attachment (bool, False): Flag for handle attachment api call. Defaults to False.
+
+        Raises:
+            InvalidResponse: An instance of InvalidResponse class.
+            Exception: An instance of an exception class.
+
+        Yields:
+            list: Table & Attachment metadata.
+            str: Attachment content.
+        """
+        retry = 1
+        offset = 0
+
+        if not is_attachment:
+            query = "ORDERBYsys_created_on^"
+            if "sysparm_query" in params.keys():
+                query += params["sysparm_query"]
+            params.update({"sysparm_query": query, "sysparm_limit": FETCH_SIZE})
+
+        while True:
+            if not is_attachment:
+                params["sysparm_offset"] = offset
+
+            try:
+                async with self._get_session().get(  # pyright: ignore
+                    url=url, params=params
+                ) as response:
+                    if (
+                        not RUNNING_FTEST
+                        and response.headers.get("Connection") == "close"
+                    ):
+                        raise Exception("Couldn't connect to ServiceNow instance")
+
+                    if is_attachment:
+                        yield response
+                        break
+
+                    fetched_response = await response.read()
+                    if fetched_response == b"":
+                        raise InvalidResponse(
+                            "Request hasn't processed from ServiceNow server"
```

Comment:
Question: does this mean that ServiceNow is still processing the request, or that they have failed to?

Reply:
Yes, request was failed and returned empty response

Reply:
In that case, I suggest:

```suggestion
                            "Request to ServiceNow server returned an empty response."
```